src/deploy/osp_deployer/deployer.py

```diff
@@ -82,16 +97,124 @@ def get_settings():
+def deploy_edge(director_vm, node_type):
+    logger.info("=== Installing edge site: %s", str(node_type))
+    director_vm.deploy_edge_site(node_type)
+
+
+def delete_edge(sah_node, director_vm, node_type):
+    logger.info("=== Deleting edge site: %s",
+                node_type)
+    director_vm.delete_edge_site(node_type)
+    logger.info("Cleaning up routes for edge site: {}".format(node_type))
+    director_vm.controller_routes_edge(node_type, False)
+    director_vm.subnet_routes_edge(node_type, False)
+    sah_node.subnet_routes_edge(node_type, False)
+
+
+def get_is_deploy_overcloud():
+    _is_deploy_overcloud = False
+    _input = input("Deploy the overcloud prior to deploying edge site(s)"
+                   ", (y/N)? ")
+    if _input.lower().strip() == "y":
+        _is_deploy_overcloud = True
+    return _is_deploy_overcloud
+
+
+def get_is_deploy_undercloud_and_overcloud():
+    _is_deploy_undercloud = False
+    _input = input("Deploy the undercloud and overcloud prior to "
+                   "deploying edge site(s), (y/N)? ")
+    if _input.lower().strip() == "y":
+        _is_deploy_undercloud = True
+    return _is_deploy_undercloud
+
+
+def deploy_undercloud(setts, sah_node, tester, director_vm):
+    director_ip = setts.director_node.public_api_ip
+    Ssh.execute_command(director_ip,
+                        "root",
+                        setts.director_node.root_password,
+                        "subscription-manager remove --all")
+    Ssh.execute_command(director_ip,
+                        "root",
+                        setts.director_node.root_password,
+                        "subscription-manager unregister")
+    sah_node.delete_director_vm()
+
+    logger.info("=== create the director vm")
+    sah_node.create_director_vm()
+    tester.director_vm_health_check()
+
+    logger.info("Preparing the Director VM")
+    director_vm.apply_internal_repos()
+
+    logger.debug(
+        "===  Uploading & configuring undercloud.conf . "
+        "environment yaml ===")
+    director_vm.upload_update_conf_files()
+
+    logger.info("=== installing the director & undercloud ===")
+    director_vm.inject_ssh_key()
+    director_vm.upload_cloud_images()
+    director_vm.install_director()
+    _is_failed, _error = tester.verify_undercloud_installed()
+    if _is_failed:
+        raise _error
+
+def list_edge_sites(settings):
+    logger.info("Available edge sites and nodes")
+    logger.info("\nEdge sites:")
+    for node_type in settings.node_types:
+        logger.info("  {}".format(node_type))
+    for node_type, node_type_data in settings.node_type_data_map.items():
+        logger.info("\nEdge site {} "
+                    "metadata:"
+                    "\n{}\n".format(node_type, pformat(node_type_data)))
+        nodes = settings.node_types_map[node_type]
+        logger.info("\nEdge site: {} nodes:".format(node_type))
+        for node in nodes:
+            logger.info("\n{}\n".format(pformat(node.__dict__)))
```

Comment:
We should clarify this is based on deployment settings, not what might currently be deployed

Reply:
+1, good idea, this brings up a good point, what if user runs deployer.py ... --edge-site edge-compute-denver twice?  Do we simply delete the stack and redeploy the edge site?
Seems reasonable, please confirm this should be the intended behavior.

I will also add deployment state to --list_edge_sites output.

Reply:
_hat if user runs deployer.py ... --edge-site edge-compute-denver twice? Do we simply delete the stack and redeploy the edge site?_
I would say yes, no different that doing an overcloud_only if overcloud already deployed, but would like to hear from @heydog15 and rest of the team also

Reply:
Done, change will be in next patch set